### PR TITLE
feat(router): new output that would notify when link is activated

### LINKED ISSUE #37284: [routerLink][routerLinkActive] Add event when link becomes active
# 🚀 feature request

### Relevant Package

 This feature request is for @angular/router

### Description
I've searched a bit but didn't find such a request or discussion which feels rather strange. Right now routerLinkActive works around change detection by applying classes directly. Therefore we cannot react to this event with something like a sliding underline in the tabs of links. It would be great if routerLinkActive directive had an output that triggers when link becomes active. It's easy to make a wrapper directive or something, but it would be nice to have it out of the box.

### Describe the solution you'd like
Just an EventEmitter<boolean> output `routerLinkActivated` on `routerLinkActive` directive.

### Describe alternatives you've considered
Create a custom directive with the same logic.

### PR DESCRIPTION
**Pull request by anandtiwary**
This commit adds a new output to `routerLinkActive` directive.
Whenever, the associated link becomes active or inactive, an
event will be fired on this out with the correct status

PR Close #37284

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #37284

## What is the new behavior?

Introduced a new Output that emits when link gets activated or inactivated

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/router/src/directives/router_link_active.ts [hunk 1]
```diff
@@ -114,35 +124,42 @@ export class RouterLinkActive implements OnChanges, OnDestroy, AfterContentInit
...
         this.update();
       }
     });
   }
 
   @Input()
-  set routerLinkActive(data: string[]|string) {
-    const classes = Array.isArray(data) ? data : data.split(' ');
-    this.classes = classes.filter(c => !!c);
+  set routerLinkActive(data: string[]|string|undefined) {
```
Comment (atscott) [inline packages/router/src/directives/router_link_active.ts:153]: Why was `undefined` added here? It doesn't seem related to this change.

File: packages/router/src/directives/router_link_active.ts [hunk 2]
```diff
@@ -99,6 +108,7 @@ export class RouterLinkActive implements OnChanges, OnDestroy, AfterContentInit
    */
   @Input() routerLinkActiveOptions: {exact: boolean}|IsActiveMatchOptions = {exact: false};
 
+  @Output() public readonly routerLinkActivated: EventEmitter<boolean> = new EventEmitter();
```
Comment (atscott) [inline packages/router/src/directives/router_link_active.ts:111]:
```suggestion
  @Output() readonly routerLinkActivated: EventEmitter<boolean> = new EventEmitter();
```
nit: `public` is not necessary

Comment (atscott) [inline packages/router/src/directives/router_link_active.ts:111]:
```suggestion
  @Output() public readonly isActiveChange: EventEmitter<boolean> = new EventEmitter();
```
Bike-shedding: I'm not sure how much I like "routerLinkActivated". How about `isActiveChange`?

File: packages/router/src/directives/router_link_active.ts [hunk 3]
```diff
@@ -163,24 +180,28 @@ export class RouterLinkActive implements OnChanges, OnDestroy, AfterContentInit
             this.renderer.removeClass(this.element.nativeElement, c);
           }
         });
+
+        // Emit on routerLinkActivated after classes are updated
+        this.routerLinkActivated.emit(hasActiveLinks);
```
Comment (atscott) [inline packages/router/src/directives/router_link_active.ts:185]: This is going to emit more frequently than we would like. It should only emit when the value changes since the last emitted value.

**Review by atscott — CHANGES_REQUESTED** [review]
I mentioned this in the previous PR - could you separate the unrelated styling changes into a different commit or PR and limit your changes to only what's necessary for this feature?

File: packages/router/src/directives/router_link_active.ts (hunk 3 above)
Comment (anandtiwary) [reply packages/router/src/directives/router_link_active.ts:185]: It is part of `if` at L173. So, won't it only emit when the two values are not matching?

File: packages/router/src/directives/router_link_active.ts (hunk 1 above)
Comment (anandtiwary) [reply packages/router/src/directives/router_link_active.ts:153]: Let me remove it

File: packages/router/src/directives/router_link_active.ts (hunk 3 above)
Comment (atscott) [reply packages/router/src/directives/router_link_active.ts:185]: Ah, you're right :)

**Comment by anandtiwary** [comment]
@atscott Some styling came from the git hook prettier. I can take care of other comments but I don't think i can revert all the formatting related changes. Let me know if you think otherwise

**Comment by atscott** [comment]
> @atscott Some styling came from the git hook prettier. 

Is this a git hook local to your machine? We use clang-format, not prettier, to format TS files.

**Comment by anandtiwary** [comment]
interesting, I do have a git hook and prettier plugin in vscode. 

If angular doesn't use prettier, how come we have `https://github.com/angular/angular/blob/master/.prettierrc` and prettier in the dev dependencies?

**Comment by atscott** [comment]
> I do see that angular has `https://github.com/angular/angular/blob/master/.prettierrc` and prettier in the dev dependencies? Are they used to format some other type of files?

That's right. We use prettier to format yaml files.

**Comment by anandtiwary** [comment]
Got it.. Thanks for your reply. This is my first PR to angular, so i am facing some tooling related issues. Let me fix those and upload a new commit.

**Comment by anandtiwary** [comment]
hmm.. So i got a build failure that branch rebase failed. So, I pulled latest master in my forked repo's master and then rebased my local `router-activated-37284` with master. And then i pushed. The logs were looking fine when i checked locally.  What did i do wrong?

**Comment by atscott** [comment]
@anandtiwary I do not know what you did with your rebase. It's probably easiest to start over again.
